tensorboard/plugins/histogram/histograms_plugin.py

```diff
@@ -173,10 +173,10 @@ def histograms_impl(self, tag, run, downsample_to=None):
               MAX(step) AS max_step
             FROM Tensors
             /* Filter out NULL so we can use TensorSeriesStepIndex. */
```

Comment:
So this comment perhaps isn't the clearest, but we specifically need to filter out null *steps* here in order to use this index, since TensorSeriesStepIndex is only defined for non-null steps.

With the `step IS NOT NULL` filter, the inner query here can do a very efficient lookup of the min and max steps using exclusively the index (it doesn't even hit the actual table).  Without it, even with a `dtype IS NOT NULL` filter, it's reduced to doing a full scan over the entire range of steps.  So for a series of n steps, this degrades performance from something likely to be O(log n) or perhaps even O(1) (if pointers are kept to the start/end of the index) to O(n).

```bash
$ sqlite3 ~/scratch/tbdemo_fuller.sqlite "EXPLAIN QUERY PLAN SELECT MIN(step), MAX(step) FROM Tensors WHERE series = 123 AND step IS NOT NULL"
0|0|0|SEARCH TABLE Tensors USING COVERING INDEX TensorSeriesStepIndex (series=?)
$ sqlite3 ~/scratch/tbdemo_fuller.sqlite "EXPLAIN QUERY PLAN SELECT MIN(step), MAX(step) FROM Tensors WHERE series = 123 AND dtype IS NOT NULL"
0|0|0|SCAN TABLE Tensors
```

The impact isn't quite as drastic in all the other cases of making this change, but in general, applying the `step IS NOT NULL` filter should produce better performance since that enables sqlite to use this index when possible.

Reply:
Ah OK. I reverted the histogram and image plugins for now. I had not thought about how the `step` is an index. At the same time, doesn't SQL query have to scan through all rows while filtering (since the `step` column is used within a `WHERE` clause)? I wonder how much `step` being an index helps there.

Reply:
This issue would be likely to affect any plugin, not just histogram/images.  I'll go into specifics below, but the TL;DR is that my point is that using the step column to look for NULL is nearly always at least as good as using dtype, and likely much more efficient in many queries.  So unless there's some really good reason to make dtype the indicator of placeholder-ness, we should use step.

SQL DB indexes pretty much exist to support WHERE clauses, so it's not really the case that WHERE clauses imply you need to do a scan... in many cases it implies the opposite.  When you use a WHERE clause filter on an indexed column where the WHERE condition is indexable (i.e. it's a pure comparison), it's more-or-less telling SQLite that it can use that index to skip ahead to values matching the filter condition instead of scanning the table.  See this section of the SQLite query optimizer: https://sqlite.org/optoverview.html#where_clause

Some specifics:

1. For the case above, SQLite can't use the TensorSeriesStepIndex at all if you don't filter for step IS NOT NULL.  That means that it actually has to scan the entire Tensors table, which is even worse than I had originally said above (it's O(# of series X max steps per series)).  With step IS NOT NULL, it can seek directly to the part of the TensorSeriesStepIndex matching the series ID provided, so it only has to look at the O(# of steps) rows for that specific series to determine the min/max.

I had thought in my above analysis that SQLite would be smart enough to use the TensorSeriesStepIndex within the series-specific range to look up the min/max in O(log # of steps) or O(1) time, but based on the full EXPLAIN results for the query, it's not that smart - it will only do that optimization when the MIN/MAX argument is the leftmost column of an index: https://sqlite.org/optoverview.html#minmax  But it's still a huge speedup to only inspect the data for a single series in this query versus all series in the Tensors table.

2. For the pr curves plugin query (and likely similar for scalars), it also makes a big difference.  The respective query plans for the PR curves query are as follows:

With `dtype IS NOT NULL`:
```text
0|0|0|SCAN TABLE Tensors
0|1|1|SEARCH TABLE Tags USING INDEX TagIdIndex (tag_id=?)
0|2|2|SEARCH TABLE Runs USING INDEX RunIdIndex (run_id=?)
```

With `step IS NOT NULL`:
```text
0|0|1|SCAN TABLE Tags USING INDEX TagRunNameIndex
0|1|2|SEARCH TABLE Runs USING INDEX RunIdIndex (run_id=?)
0|2|0|SEARCH TABLE Tensors USING INDEX TensorSeriesStepIndex (series=?)
```

Again, we can only use TensorSeriesStepIndex if we filter out null steps.  SQLite joins are implemented as nested for loops, so in the first case, we are literally looping over the entire tensors table again (even though the query is targeting a specific set of runs and a specific single tag).  For each Tensor we have to join it to its tag and run in order to check if they match the tag and run name.  In the second case, we instead scan TagRunNameIndex which is O(# of unique run-tag combinations), which reduces the scanned surface by O(# of steps per tag) like above.  We then do an index lookup into Runs to check the run_name, and then use the tag_id to index into Tensors to scan data for just that tag.

This could actually be improved further by adding `AND experiment_id IS NULL` to the query, which then lets us use RunNameIndex to look up the run_id, which in turns lets us use TagRunNameIndex to do an index lookup into Tags instead of a scan.  The resulting query plan is:

```text
0|0|0|SEARCH TABLE Runs USING INDEX RunNameIndex (experiment_id=? AND run_name=?)
0|1|1|SEARCH TABLE Tags USING INDEX TagRunNameIndex (run_id=? AND tag_name=?)
0|2|2|SEARCH TABLE Tensors USING INDEX TensorSeriesStepIndex (series=?)
```

This is another drastic improvement since we now are doing work per run queried that's O( log( # of runs) + log(# of tags) + log(# of tensors) + # of tensors for this specific tag), so it scales linearly only with the actual data that we're returning instead of the full DB size.